PD-440 CLI Disk Namespace (#2077)
truenas/documentation: content/SCALECLIReference/Storage/CLIDisk.md

|       critical | <null>                                  |
|     difference | <null>                                  |
|  informational | <null>                                  |
|          model | HUS726020AL4210                         |
|   rotationrate | 7200                                    |
|           type | HDD                                     |
|       zfs_guid | 5249139309134546836                     |
|            bus | SCSI                                    |
|        devname | sdl                                     |
|      enclosure | <dict>                                  |
| supports_smart | <null>                                  |
|           pool | <null>                                  |
+----------------+-----------------------------------------+
```
{{< /expand >}}
{{< /expand >}}

### Get_Unused Command

The `get_unused` command returns disks that are not in use by any zpool. It will also return disks that are in use by any zpool that is exported. You can add the optional `join_partitions` argument to list disk partitions as well.

{{< expand "Listing Unused Disks" "v" >}}

#### Description

> The `get_unused` command does not require entering properties or arguments, but it does contain the optional `join_partitions` argument.

Reviewer: ```suggestion
The `get_unused` command does not have required properties , but it does have one optional property  `join_partitions` .
```
Below this add something like:
`join_partitions` has xxx properties. 
Or you can just tell the reader to go see the next section on using this property (which I just saw). 
Then enter the instruction for how to enter the property argument.

| {serial_lunid}N4G22WKK_5000cca24503ce58     | sdl  | scsi      | 2224   | N4G22WKK     | 5000cca24503ce58 | 2000398934016 |             | Auto         | Always On  | Disabled     | true        |              | <null>     |        | <null>   | <null>     | <null>        | HUS726020AL4210           | 7200         | HDD  | <null>   | <null>   | SCSI | <null>         | <null>                 | <empty list>     | <dict>    | <empty list> | sdl     |
+---------------------------------------------+------+-----------+--------+--------------+------------------+---------------+-------------+--------------+------------+--------------+-------------+--------------+------------+--------+----------+------------+---------------+---------------------------+--------------+------+----------+----------+------+----------------+------------------------+------------------+-----------+--------------+---------+
```
{{< /expand >}}
{{< /expand >}}

{{< expand "Listing Unused Disks with Partitions" "v" >}}

#### Description

> The `join_partitions` argument determines if the `get_unused` command returns all partitions written to each disk. The default is false.

Reviewer: After this, tell them how to enter the argument. Also in this we should call it the property.
```suggestion
The `join_partitions` property argument determines if the `get_unused` command returns all partitions written to each disk. The default is false.
```

| {serial_lunid}N4G22WKK_5000cca24503ce58     | sdl  | scsi      | 2224   | N4G22WKK     | 5000cca24503ce58 | 2000398934016 |             | Auto         | Always On  | Disabled     | true        |              | <null>     |        | <null>   | <null>     | <null>        | HUS726020AL4210           | 7200         | HDD  | <null>   | <null>              | SCSI | <null>         | <null>                 | <empty list>     | <dict>    | <list>       | sdl     |
+---------------------------------------------+------+-----------+--------+--------------+------------------+---------------+-------------+--------------+------------+--------------+-------------+--------------+------------+--------+----------+------------+---------------+---------------------------+--------------+------+----------+---------------------+------+----------------+------------------------+------------------+-----------+--------------+---------+
```
{{< /expand >}}
{{< /expand >}}

### Query Command

The `query` command displays disk information.

{{< expand "Running a Simple Disk Query" "v" >}}

#### Description

> Enter the  `query` command with no additional attributes to perform a basic query of all disks.

Reviewer: ```suggestion
The  `query` command does not require entering properties or arguments to perform a basic query of all disks.
```

```
storage disk query name
+---------+
| name    |
+---------+
| sdb     |
| sda     |
| nvme0n1 |
| sdc     |
| sdd     |
| sdm     |
| sdn     |
| sde     |
| sdf     |
| sdg     |
| sdh     |
| sdi     |
| sdj     |
| sdk     |
| sdl     |
+---------+
```
{{< /expand >}}
{{< /expand >}}

### Resize Command

The `resize` command allows you to resize disks to a specific amount of gigabytes.

{{< expand "Resizing Disks">}}

#### Description
The `resize` command has three properties. They are `disks`, `sync`, and `raise_error`.
Enter the command string, then press <kbd>Enter</kbd>.
The command returns completion percentages when successful.

#### Usage

From the CLI prompt, enter:

<code>storage disk resize disks={"name":"<i>diskname</i>,<i>diskname</i>", "size":"<i>number</i>"} sync=<i>true/false</i> raise_error=<i>true/false</i><code>

Where:
* *diskname* is the name of the disk you want to resize. Separate multiple disks with commas
* *number* is the size (in gigabytes) you want to resize the disk to.
* *true/false* in the `sync` property enables or disables synchronizing the new size of the disk(s) with the database cache. The default is true.

> * *true/false* in the `raise_error` property sets whether the disk(s) raise a CallError upon failure (true) or will only log errors (false). The default is false.

Reviewer: do not use will, and this detailed explanation goes in in a table describing the property and how to enter the argument. The Where section is about explaining the variable not the arguments.

[0%] ...
[100%] ...
```
{{< /expand >}}
{{< /expand >}}

### Retaste Command

The `retaste` command forces the system to re-read specified disks and update their data.

{{< expand "Retasting Disks">}}

#### Description
The `retaste` command requires you to specify which disks you want to retaste.
Enter the command string, then press <kbd>Enter</kbd>.
The command returns completion percentages when successful.

#### Usage

From the CLI prompt, enter:

<code>storage disk retaste disks=<i>name</i>,<i>name</i><code>

Where *name* is the name of a disk you want to retaste. Separate each disk name with a comma.

{{< expand "Command Example" "v" >}}
````
storage disk retaste disks=sdl
[85%] Retasting disks...
[95%] Waiting for disk events to settle...
[100%] Retasting disks done...
SUCCESS
````
{{< /expand >}}
{{< /expand >}}

### Smart_Attributes Command

The `smart_attributes` command returns S.M.A.R.T. attributes values for specified disk.

{{< hint type=note >}}
Only devices with the ATA bus type support S.M.A.R.T. attributes.
{{< /hint >}}

{{< expand "Viewing S.M.A.R.T. Attributes">}}

#### Description

> The `smart_attributes` command requires you to specify which disk you want to view S.M.A.R.T. attributes for.

Reviewer: Use the standard instruction here such as:
```suggestion
The `smart_attributes` command has one required property, `name` to specify which disk you want to view S.M.A.R.T. attributes for.
```
also, is name basically the type of disk?

> Enter the command string, then press <kbd>Enter</kbd>.

Reviewer: Tell them how to enter the property argument (w/wo double-quoted values if that apples). Follow the how with this sentence.

| 197 | Current_Pending_Sector  | 100   | 100   | 0      |             | <dict> | <dict> |
| 198 | Offline_Uncorrectable   | 100   | 100   | 0      |             | <dict> | <dict> |
| 199 | UDMA_CRC_Error_Count    | 100   | 100   | 0      |             | <dict> | <dict> |
| 202 | Percent_Lifetime_Remain | 100   | 100   | 1      |             | <dict> | <dict> |
| 206 | Write_Error_Rate        | 100   | 100   | 0      |             | <dict> | <dict> |
| 246 | Total_LBAs_Written      | 100   | 100   | 0      |             | <dict> | <dict> |
| 247 | Host_Program_Page_Count | 100   | 100   | 0      |             | <dict> | <dict> |
| 248 | Bckgnd_Program_Page_Cnt | 100   | 100   | 0      |             | <dict> | <dict> |
| 180 | Unused_Rsvd_Blk_Cnt_Tot | 1     | 1     | 0      |             | <dict> | <dict> |
| 210 | RAIN_Success_Recovered  | 100   | 100   | 0      |             | <dict> | <dict> |
+-----+-------------------------+-------+-------+--------+-------------+--------+--------+
```
{{< /expand >}}
{{< /expand >}}

### Temperature Command

The `temperature` command returns the temperature of a specified disk.

{{< expand "Checking Disk Temperature">}}

#### Description

> The `temperature` command requires you to specify which disk you want to see the temperature for. You can also set the S.M.A.R.T. `powermode`, and retrieve previous temperatures using the `cache` option.

Reviewer: What is the required property? is it `name` or some other? Include how to enter the property argument.
```suggestion
The `temperature` command has one required property, `name` to specify which disk you want to see the temperature for. 
`temperature` has two optional properties, `powermode` and `cache`.
Use the optional S.M.A.R.T. property `powermode`  to retrieve ????.
Use `cahce` to retrieve previous temperatures.
```
Not sure what `powermode` does for the user if they enter it. Create a small table for these or just describe in a sentence, and include the syntax for the arguments for each if not the standard text for entering a property argument.

Enter the command string, then press <kbd>Enter</kbd>.
The command returns a temperature (in Celsius) when successful.

#### Usage

From the CLI prompt, enter:

<code>storage disk temperature name=<i>diskname</i> options={"cache":"<i>seconds</i>","powermode":"<i>MODE</i>"}<code>

Where 
* *diskname* is the name of a disk.

> * *seconds* is how far back in seconds you want to view the disk temperature. The default is null and returns the current temperature.

Reviewer: the Null sentence goes with the description of the option. In this case `options` is the property and it has two required options `cache` and `powermode`. My comments above should reflect this but I didn't know how to use them when I reviewed that text.

|-----------|-----------------------------------------------------------------------------------------------------------------------------------------|
| `NEVER`   | The device is fully powered up and ready to send/receive data. The disk only undergoes S.M.A.R.T. tests when powermode is set to NEVER. |
| `IDLE`    | Disk completes commands slower than when set to NEVER but uses less power.                                                             |
| `STANDBY` | Disk completes commands slower than when set to IDLE but uses less power.                                                              |
| `SLEEP`   | Disk does not complete commands until reset. Uses the least amount of power
{{< /truetable >}}                                                             |
{{< /expand >}}

{{< expand "Command Example" "v" >}}
```
storage disk temperature name=sda options={"cache":"30","powermode":"NEVER"}
40
```
{{< /expand >}}
{{< /expand >}}

### Temperature_Agg Command

The `temperature_agg` command returns min/max/avg temperature for specified disks for a set amount of previous days.

{{< expand "Checking Disk Temperatures Over Time">}}

#### Description

> The `temperature_agg` command requires you to specify which disks you want to see the temperatures for and the number of days you wish to include.

Reviewer: same comments as above.

Where 
* *diskname* is the name of a disk.
* *number* is the number of days you want to view temperatures for.

{{< expand "Command Example" "v" >}}
```
storage disk temperature_agg names=sda,sdb days=5
+-----+--------+
| sdb | <dict> |
| sda | <dict> |
+-----+--------+
```
{{< /expand >}}
{{< /expand >}}

### Temperature_Alerts Command

The `temperature_alerts` command returns existing temperature alerts for specified disks.

{{< expand "Viewing Temperature Alerts" "v" >}}
#### Description
The `temperature_alerts` command requires you to specify the names of each disk for which you want to see existing temperature alerts.
Enter the command string, then press <kbd>Enter</kbd>.
The command returns a list with all existing disk alerts.

#### Usage

From the CLI prompt, enter:

<code>storage disk temperature_alerts names=<i>diskname</i>,<i>diskname</i>

Where *diskname* is the name of a disk you want to view alerts for. Separate each disk name with a comma.

{{< expand "Command Example" "v" >}}
```
storage disk temperature_alerts names=sda,sdb
(empty list)
```
{{< /expand >}}
{{< /expand >}}

> ### Temperatures Command

Reviewer: I think this is a duplicate section to the Temperature command above.

> The `temperatures` command

Reviewer: I think you have an unwanted CR here.

|-----------|-----------------------------------------------------------------------------------------------------------------------------------------|
| `NEVER`   | The device is fully powered up and ready to send/receive data. The disk only undergoes S.M.A.R.T. tests when powermode is set to NEVER. |
| `IDLE`    | Disk completes commands slower than when set to NEVER, but uses less power.                                                             |
| `STANDBY` | Disk completes commands slower than when set to IDLE, but uses less power.                                                              |
| `SLEEP`   | Disk does not complete commands until reset. Uses the least amount of power
{{< /truetable >}}                                                             |
{{< /expand >}}

{{< expand "Command Example" "v" >}}
```
storage disk temperatures name=sda,sdb options=NEVER
+-----+----+
| sda | 40 |
| sdb | 40 |
+-----+----+
```
{{< /expand >}}
{{< /expand >}}

### Update Command

The `update` command allows you to update settings for a specified disk.

{{< expand "Updating Disks">}}

#### Description

> The `update` command has 15 configurable properties. After specifying the `id` of the disk you want to update, you must include at least one property to update.

Reviewer: same comments as above.
